src/PrestaShopBundle/Translation/PrestaShopTranslatorTrait.php

```diff
@@ -53,28 +53,22 @@ public function trans($id, array $parameters = [], $domain = null, $locale = nul
             $legacy = $parameters['legacy'];
             unset($parameters['legacy']);
         }
+        $isSprintf = !empty($parameters) && $this->isSprintfString($id);
 
         if (empty($locale)) {
             $locale = null;
         }
 
-        $translated = parent::trans($id, [], $this->normalizeDomain($domain), $locale);
-
-        // @todo to remove after the legacy translation system has ben phased out
-        if ($this->shouldFallbackToLegacyModuleTranslation($id, $domain, $translated)) {
+        if ($this->shouldFallbackToLegacyModuleTranslation($id, $domain)) {
             return $this->translateUsingLegacySystem($id, $parameters, $domain, $locale);
         }
 
-        if (isset($legacy) && 'htmlspecialchars' === $legacy) {
-            $translated = call_user_func($legacy, $translated, ENT_NOQUOTES);
-        } elseif (isset($legacy)) {
-            $translated = call_user_func($legacy, $translated);
-        }
+        $translated = parent::trans($id, $isSprintf ? [] : $parameters, $this->normalizeDomain($domain), $locale);
 
-        if (!empty($parameters) && $this->isSprintfString($id)) {
+        $translated = isset($legacy) ? $this->replaceSpecialCharsWithLegacyFunctions($translated, $legacy) : $translated;
```

Comment:
Wouldn't it make more sense to replace the special characters after the tokens have been replaced instead of doing it before?

Reply:
Not sure what difference does it make technically. But I can see how more "natural" flow it is.  :)

Reply:
Here's why.

Imagine this scenario:

```php
trans("Some <b>%s</b>", ["<i>injection</i>", 'legacy' => 'htmlspecialchars'], 'foo');
```

If you do the replacement before, you get:

```
Some &lt;<i>injection</i>&gt;
```

If you the replacement after, you get:

```
Some &lt;&lt;i&gt;injection&lt;/i&gt;&gt;
```